grpc/optimistic/server.go

```diff
@@ -0,0 +1,308 @@
+	auctionService := &AuctionServiceV1Alpha1{
+		sharedServiceContainer: sharedServiceContainer,
+	}
+
+	auctionService.currentAuctionBlock.Store(&[]byte{})
+
+	return auctionService
+}
+
+func (o *AuctionServiceV1Alpha1) GetBidStream(_ *auctionPb.GetBidStreamRequest, stream auctionGrpc.AuctionService_GetBidStreamServer) error {
+	log.Debug("GetBidStream called")
+
+	pendingTxEventCh := make(chan core.NewTxsEvent)
+	pendingTxEvent := o.Eth().TxPool().SubscribeTransactions(pendingTxEventCh, false)
+	defer pendingTxEvent.Unsubscribe()
+
+	for {
+		select {
+		case pendingTxs := <-pendingTxEventCh:
+			// get the optimistic block
+			// this is an in-memory read, so there shouldn't be a lot of concerns on speed
+			optimisticBlock := o.Eth().BlockChain().CurrentOptimisticBlock()
+
+			for _, pendingTx := range pendingTxs.Txs {
+				bid := auctionPb.Bid{}
+
+				totalCost := big.NewInt(0)
+				effectiveTip := cmath.BigMin(pendingTx.GasTipCap(), new(big.Int).Sub(pendingTx.GasFeeCap(), optimisticBlock.BaseFee))
+				totalCost = totalCost.Mul(effectiveTip, big.NewInt(int64(pendingTx.Gas())))
+
+				marshalledTxs := [][]byte{}
+				marshalledTx, err := pendingTx.MarshalBinary()
+				if err != nil {
+					return status.Errorf(codes.Internal, shared.WrapError(err, "error marshalling tx").Error())
+				}
+				marshalledTxs = append(marshalledTxs, marshalledTx)
+
+				bid.Fee = totalCost.Uint64()
+				bid.Transactions = marshalledTxs
+				bid.SequencerParentBlockHash = *o.currentAuctionBlock.Load()
+				bid.RollupParentBlockHash = optimisticBlock.Hash().Bytes()
+
+				txsStreamedCount.Inc(1)
+				err = stream.Send(&auctionPb.GetBidStreamResponse{Bid: &bid})
+				if err != nil {
+					log.Error("error sending bid over stream", "err", err)
+					return status.Error(codes.Internal, shared.WrapError(err, "error sending bid over stream").Error())
+				}
+			}
+
+		case err := <-pendingTxEvent.Err():
+			if err != nil {
+				log.Error("error waiting for pending transactions", "err", err)
+				return status.Error(codes.Internal, shared.WrapError(err, "error waiting for pending transactions").Error())
+			} else {
+				// TODO - what is the right error code here?
+				return status.Error(codes.Internal, "tx pool subscription closed")
+			}
+
+		case <-stream.Context().Done():
+			return stream.Context().Err()
+		}
+	}
+}
+
+func (o *AuctionServiceV1Alpha1) ExecuteOptimisticBlockStream(stream auctionGrpc.OptimisticExecutionService_ExecuteOptimisticBlockStreamServer) error {
+	log.Debug("ExecuteOptimisticBlockStream called")
+
+	mempoolClearingEventCh := make(chan core.NewMempoolCleared)
+	mempoolClearingEvent := o.Eth().TxPool().SubscribeMempoolClearance(mempoolClearingEventCh)
+	defer mempoolClearingEvent.Unsubscribe()
+
+	for {
+		msg, err := stream.Recv()
+		// stream has been closed
+		if errors.Is(err, io.EOF) {
+			return nil
+		}
+		if err != nil {
+			return err
```

Comment:
wrap in `status.Error`?

Reply:
addressed in https://github.com/astriaorg/flame/pull/30/commits/defc35bb0c32331a823a635dccdb28c7e62d519b